File: packages/contracts/adapters.ts
import {
  CostLatencySchema,
  EvidenceManifestSchema,
  PipelineEventSchema,
  TraceContextSchema,
} from "./generated/typescript/index.js";

type WirePayload = Record<string, unknown>;

export type PipelineMetadata = {
  trace_context?: unknown;
  cost_latency?: unknown;
  evidence_manifest?: unknown;
};

/**
 * Add optional pipeline metadata without renaming, filtering, or coercing the
 * existing payload. This keeps legacy HTTP wire shapes contract-compatible.
 */
export function attachPipelineMetadata<T extends WirePayload>(
  payload: T,
  metadata: PipelineMetadata,
): T & PipelineMetadata {
  return { ...payload, ...metadata };
}

/** Validate an additive event envelope while preserving the established event vocabulary. */
export function createPipelineEvent(event: WirePayload): WirePayload {
  return PipelineEventSchema.parse(event);
}

export const pipelineMetadataSchemas = {
  trace_context: TraceContextSchema,
  cost_latency: CostLatencySchema,
  evidence_manifest: EvidenceManifestSchema,
};

File: packages/contracts/adapters.py
"""Compatibility adapters for adding pipeline metadata to existing wire payloads."""

from collections.abc import Mapping
from typing import Any


def attach_pipeline_metadata(
    payload: Mapping[str, Any],
    *,
    trace_context: Mapping[str, Any] | None = None,
    cost_latency: Mapping[str, Any] | None = None,
    evidence_manifest: Mapping[str, Any] | None = None,
) -> dict[str, Any]:
    """Return a copy with additive metadata and every legacy field unchanged."""
    adapted = dict(payload)
    if trace_context is not None:
        adapted["trace_context"] = dict(trace_context)
    if cost_latency is not None:
        adapted["cost_latency"] = dict(cost_latency)
    if evidence_manifest is not None:
        adapted["evidence_manifest"] = dict(evidence_manifest)
    return adapted
